source/postprocess/visualization/seismic_melt.cc

```diff
@@ -0,0 +1,372 @@
+        MaterialModel::MeltOutputs<dim> *melt_outputs = out.template get_additional_output<MaterialModel::MeltOutputs<dim> >();
+        AssertThrow(melt_outputs != NULL,
+                    ExcMessage("Need MeltOutputs from the material model for computing the melt properties."));
+
+        const unsigned int porosity_idx = this->introspection().compositional_index_for_name("porosity");
+
+        for (unsigned int q=0; q<n_quadrature_points; ++q)
+          {
+            unsigned output_index = 0;
+            for (unsigned int i=0; i<property_names.size(); ++i, ++output_index)
+              {
+                double porosity = std::max(in.composition[q][porosity_idx],0.0);
+
+                if (porosity < 0)
+                  porosity = 0.0;
+
+                double Vp = 0.0;
+                double Vs = 0.0;
+                double contiguity = 0.0;
+
+                const double solid_bulk_modulus = solid_bulk_modulus_GPa * std::pow(10,9);
+                const double liquid_bulk_modulus = liquid_bulk_modulus_GPa * std::pow(10,9);
+                const double shear_modulus = shear_modulus_GPa * std::pow(10,9);
+
+                double n_k = 0.0;
+                double n_mu = 0.0;
+
+                double solid_bulk_modulus_sk = 0.0;
+                double shear_modulus_sk = 0.0;
+
+                double skeleton_solid_bulk_modulus = 0.0;
+                double skeleton_shear_modulus = 0.0;
+
+                double a_1 = 0.0;
+                double a_2 = 0.0;
+                double a_3 = 0.0;
+                double b_1 = 0.0;
+                double b_2 = 0.0;
+
+                if (porosity > 0)
+                  {
+                    //contiguity relation from Wimert and Hier-Majumder 2011
+                    contiguity = p1 * std::pow(porosity,5) + p2 * std::pow(porosity,4) + p3 * std::pow(porosity,3) + p4 * std::pow(porosity,2) + p5 * porosity + p6 ;
+                    if (contiguity < 0)
+                      contiguity = 0.0;
+
+                    // following from Takei (1998), Takei (2000), and Takei (2001)
+                    const double a_hat [3][4] =
+                    {
+                      { 0.318, 6.780, 57.560, 0.182 },
+                      { 0.164, 4.290, 26.658, 0.464 },
+                      { 1.549, 4.814, 8.777, -0.290 }
+                    };
+
+                    const double b_hat [2][2] =
+                    {
+                      { -0.3238, 0.2341},
+                      { -0.1819, 0.5103},
+                    };
+
+                    a_1 = a_hat[0][0] * std::exp(((a_hat[0][1] * (poisson - 0.25)) + a_hat[0][2] * std::pow((poisson - 0.25),3))) + a_hat[0][3] ;
+                    a_2 = a_hat[1][0] * std::exp(((a_hat[1][1] * (poisson - 0.25)) + a_hat[1][2] * std::pow((poisson - 0.25),3))) + a_hat[1][3] ;
+                    a_3 = a_hat[2][0] * std::exp(((a_hat[2][1] * (poisson - 0.25)) + a_hat[2][2] * std::pow((poisson - 0.25),3))) + a_hat[2][3] ;
```

Comment:
It is a bit more elegant and shorter if you write this as a loop from 1...3 instead of copying the expression three times.